File: client/src/main/java/org/evosuite/kex/KexService.java
package org.evosuite.kex;

import org.evosuite.Properties;
import org.evosuite.classpath.ClassPathHandler;
import org.evosuite.testcase.DefaultTestCase;
import org.evosuite.testcase.execution.ExecutionObserver;
import org.evosuite.testcase.execution.ExecutionResult;
import org.evosuite.testcase.execution.TestCaseExecutor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.vorpal.research.kex.config.FileConfig;
import org.vorpal.research.kex.config.GlobalConfig;
import org.vorpal.research.kex.config.GlobalConfigKt;
import org.vorpal.research.kex.config.RuntimeConfig;
import org.vorpal.research.kex.launcher.ConcolicLauncher;
import org.vorpal.research.kex.util.RtKt;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.Arrays;
import java.util.List;
import java.util.Set;

public class KexService {

    private static final Logger logger = LoggerFactory.getLogger(KexService.class);

    private static KexService instance = null;

    private final ConcolicLauncher launcher;

    private final KexClassLoader loader;

    private KexService(List<String> classPaths, String targetName, URL[] instrumented) {
        launcher = new ConcolicLauncher(classPaths, targetName);
        loader = new KexClassLoader(instrumented);
    }

    public static KexService getInstance() {
        if (instance == null) {
            GlobalConfig kexConfig = GlobalConfigKt.getKexConfig();
            kexConfig.initialize(RuntimeConfig.INSTANCE, new FileConfig("kex.ini"));
            URL instrumentedDir;
            try {
                instrumentedDir = RtKt.getInstrumentedCodeDirectory(kexConfig).toUri().toURL();
            } catch (MalformedURLException e) {
                throw new RuntimeException(e);
            }

            String[] classPaths = ClassPathHandler.getInstance().getClassPathElementsForTargetProject();
            instance = new KexService(Arrays.asList(classPaths), Properties.TARGET_CLASS, new URL[]{instrumentedDir});
        }
        return instance;
    }

    public void execute(DefaultTestCase defaultTestCase) {
        // Kex preparation
        KexObserver kexObserver = new KexObserver(launcher.getContext());

        // Evosuite preparation
        defaultTestCase.changeClassLoader(loader);

        // Execution preparation
        Set<ExecutionObserver> originalExecutionObservers = TestCaseExecutor.getInstance().getExecutionObservers();
        TestCaseExecutor.getInstance().newObservers();
        TestCaseExecutor.getInstance().addObserver(kexObserver);

        // Execution
        ExecutionResult result = null;
        try {
            result = TestCaseExecutor.getInstance().execute(defaultTestCase, Properties.CONCOLIC_TIMEOUT);
        } catch (Exception e) {
            logger.error("Exception during kex execution: ", e);
        } finally {
            TestCaseExecutor.getInstance().setExecutionObservers(originalExecutionObservers);
        }
    }

}

File: client/src/main/kotlin/org/evosuite/kex/KexService.kt
package org.evosuite.kex

import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.InternalSerializationApi
import org.evosuite.Properties
import org.evosuite.classpath.ClassPathHacker
import org.evosuite.classpath.ClassPathHandler
import org.evosuite.kex.observers.KexObserver
import org.evosuite.testcase.DefaultTestCase
import org.evosuite.testcase.execution.ExecutionResult
import org.evosuite.testcase.execution.TestCaseExecutor
import org.slf4j.LoggerFactory
import org.vorpal.research.kex.compile.JavaCompilerDriver
import org.vorpal.research.kex.config.FileConfig
import org.vorpal.research.kex.config.RuntimeConfig
import org.vorpal.research.kex.config.kexConfig
import org.vorpal.research.kex.launcher.ConcolicLauncher
import org.vorpal.research.kex.reanimator.codegen.javagen.ReflectionUtilsPrinter
import org.vorpal.research.kex.util.asmString
import org.vorpal.research.kex.util.compiledCodeDirectory
import org.vorpal.research.kex.util.instrumentedCodeDirectory
import org.vorpal.research.kex.util.testcaseDirectory
import org.vorpal.research.kfg.ir.Method
import java.io.File
import kotlin.io.path.absolutePathString
import kotlin.io.path.div
import kotlin.time.ExperimentalTime


@ExperimentalTime
@InternalSerializationApi
@ExperimentalSerializationApi
@DelicateCoroutinesApi
object KexService {
    private val logger = LoggerFactory.getLogger(KexService::class.java)

    private lateinit var launcher: ConcolicLauncher
    private lateinit var loader: KexClassLoader

    lateinit var reflectionUtils: ReflectionUtilsPrinter

    val ctx get() = launcher.context
    val fakeEmptyMethod: Method
        get() = ctx.cm[KexService::class.java.name].getMethod("fakeEmptyMethod", ctx.types.voidType)

    @JvmStatic
    @Suppress("UNUSED")
    fun fakeEmptyMethod() {
    }

    @JvmStatic
    fun init() {
        kexConfig.initialize(RuntimeConfig, FileConfig("kex.ini"))

        val instrumentedDir = kexConfig.instrumentedCodeDirectory.toUri().toURL()
        val classPaths = ClassPathHandler.getInstance().classPathElementsForTargetProject.asList()

        launcher = ConcolicLauncher(classPaths, Properties.TARGET_CLASS)
        loader = KexClassLoader(arrayOf(instrumentedDir))

        initReflectionUtils()
    }

    private fun initReflectionUtils() {
        val pack = Properties.TARGET_CLASS.substringBeforeLast('.', "")
        reflectionUtils = ReflectionUtilsPrinter.reflectionUtils(pack)
        val compileDir = kexConfig.compiledCodeDirectory.also {
            it.toFile().mkdirs()
        }
        val reflectionFile =
            kexConfig.testcaseDirectory / pack.asmString / "${ReflectionUtilsPrinter.REFLECTION_UTILS_CLASS}.java"
        val driver = JavaCompilerDriver(emptyList(), compileDir)
        driver.compile(listOf(reflectionFile))

        ClassPathHacker.addURL(compileDir.toUri().toURL())

        val cp = ClassPathHandler.getInstance().evoSuiteClassPath.split(File.pathSeparator).toTypedArray()
        ClassPathHandler.getInstance().setEvoSuiteClassPath(cp + compileDir.absolutePathString())
    }

    @JvmStatic
    fun execute(
        defaultTestCase: DefaultTestCase,
        kexObserver: KexObserver
    ): ExecutionResult? {
        defaultTestCase.changeClassLoader(loader)

        val originalExecutionObservers = TestCaseExecutor.getInstance().executionObservers
        TestCaseExecutor.getInstance().newObservers()
        TestCaseExecutor.getInstance().addObserver(kexObserver)

        // Execution
        return try {
            TestCaseExecutor.getInstance().execute(defaultTestCase, Properties.CONCOLIC_TIMEOUT)
        } catch (e: Exception) {
            logger.error("Exception during kex execution: ", e)
            null
        } finally {
            TestCaseExecutor.getInstance().executionObservers = originalExecutionObservers
        }
    }

}
